api/admin/controller.py

```diff
@@ -2323,6 +2325,107 @@ def _sitewide_settings_controller(self, configuration_object):
+        protocols.append(dict(name=ExternalIntegration.MANUAL,
+                              label=_("Manual import"),
+                              description=_("Books will be manually added to the circulation manager, not imported automatically through a protocol."),
+                              settings=[],
+                              ))
+
+        return protocols
+
+    def _get_prior_test_results(self, collection, protocolClass):
+        """This helper function returns previous self test results for a given
+        collection if it has a protocol.
+        """
+        provider_apis = [OPDSImportMonitor,
+                         OPDSForDistributorsAPI,
+                         OverdriveAPI,
+                         OdiloAPI,
+                         BibliothecaAPI,
+                         Axis360API,
+                         OneClickAPI,
+                         EnkiAPI,
+                         ODLWithConsolidatedCopiesAPI,
+                         SharedODLAPI,
+                         FeedbooksOPDSImporter,
+                        ]
+
+        self_test_results = None
+        protocol = protocolClass
+
+        if not collection or not collection.protocol:
+            return None
+
+        if collection.protocol == OPDSImportMonitor.PROTOCOL:
+            protocol = OPDSImportMonitor
+
+        if protocol in provider_apis and issubclass(protocol, HasSelfTests):
+            if (collection.protocol == OPDSImportMonitor.PROTOCOL):
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection, OPDSImporter)
+            else:
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection)
+
+        return self_test_results
+
+    def collection_self_tests(self, identifier):
+        provider_apis = [OPDSImporter,
+                         OPDSForDistributorsAPI,
+                         OverdriveAPI,
+                         OdiloAPI,
+                         BibliothecaAPI,
+                         Axis360API,
+                         OneClickAPI,
+                         EnkiAPI,
+                         ODLWithConsolidatedCopiesAPI,
+                         SharedODLAPI,
+                         FeedbooksOPDSImporter,
+                        ]
+        protocols = self._get_protocols(provider_apis)
+
+        if not identifier:
+            return Response("No Identifier", 200)
+
+        collection = dict()
+        collectionFound = None
+        protocolClass = None
+        for col in self._db.query(Collection).filter(Collection.id==int(identifier)):
+            collectionFound = col
+            collection = dict(
+                id=col.id,
+                name=col.name,
+                protocol=col.protocol,
+                parent_id=col.parent_id,
+                settings=dict(external_account_id=col.external_account_id),
+            )
```

Comment:
You can build the `collection` object inside `if flask.request.method == 'GET'`. I think that will save you from having to mock `self._get_prior_test_results` in some of the tests. Not a big deal though.

Reply:
Updated. That was helpful to not have to mock some functions.